### PR TITLE
Assignment9+10: With Jessica Ho

### PR DESCRIPTION
**Pull request by jacobnewcomb**
@JessicaH46

### CODE REVIEW

**Comment by Afterlyte** [comment]
I will review this

File: Assignment/Assignment.Tests/Assignment.Tests.csproj [hunk 1]
```diff
@@ -1,6 +1,9 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
+	<TargetFramework>net7.0</TargetFramework>
```
Comment (Afterlyte) [inline Assignment/Assignment.Tests/Assignment.Tests.csproj:4]: Shared Properties like TargetFramework and Nullable shouldn't be here if they're in Directory.Build.Props already

File: Assignment/Assignment.Tests/PingProcessTests.cs [hunk 2]
```diff
@@ -57,44 +58,42 @@ public void Run_CaptureStdOutput_Success()
...
     }
 
     [TestMethod]
     public void RunAsync_UsingTaskReturn_Success()
     {
-        // Do NOT use async/await in this test.
-        PingResult result = default;
-        // Test Sut.RunAsync("localhost");
-        AssertValidPingOutput(result);
+        AssertValidPingOutput(Sut.RunAsync("localhost").Result);
```
Comment (Afterlyte) [inline Assignment/Assignment.Tests/PingProcessTests.cs:68]: It's a bit more readable if you separate function calls into variables (like here, having RunAsync return as a PingResult object) so it's clear what you're trying to test with at first glance.

```suggestion
        PingResult result = Sut.RunAsync("localhost").Result);
        
        AssertValidPingOutput(result);

```

File: Assignment/Assignment/PingProcess.cs [hunk 3]
```diff
@@ -27,26 +27,29 @@ public PingResult Run(string hostNameOrAddress)
...
         string hostNameOrAddress, CancellationToken cancellationToken = default)
     {
-        Task task = null!;
-        await task;
-        throw new NotImplementedException();
+        while (true)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            return await Task.Run(() => RunTaskAsync(hostNameOrAddress), cancellationToken);
+        }
```
Comment (Afterlyte) [inline Assignment/Assignment/PingProcess.cs:40]: I'm very confused about this while loop, and it feels very risky to just have a "while true" loop in here.
…Although all the tests concerning it passed without issue so maybe I just don't quite understand it.

File: Assignment/Assignment/PingProcess.cs [hunk 4]
```diff
@@ -58,9 +61,9 @@ async public Task<PingResult> RunAsync(params string[] hostNameOrAddresses)
     async public Task<PingResult> RunLongRunningAsync(
         string hostNameOrAddress, CancellationToken cancellationToken = default)
     {
-        Task task = null!;
-        await task;
-        throw new NotImplementedException();
+        Task<PingResult> task = Task.Factory.StartNew(() =>
+            Run(hostNameOrAddress), cancellationToken, TaskCreationOptions.LongRunning, TaskScheduler.Current);
```
Comment (Afterlyte) [inline Assignment/Assignment/PingProcess.cs:65]: Requirements specified to not use the Task.Run method. (Hint: look at how the Run method is implemented to figure out how to do it without using it explicitly)

**Review by Afterlyte — COMMENTED** [review]
1. Implement `PingProcess`' `public Task<PingResult> RunTaskAsync(string hostNameOrAddress)` ✔
   - First implement `public void RunTaskAsync_Success()` test method to test `PingProcess.RunTaskAsync()` using `"localhost"`. ✔
   - Do NOT use async/await in this implementation. ✔
2. Implement `PingProcess`' `async public Task<PingResult> RunAsync(string hostNameOrAddress)` ✔
   - First implement the `public void RunAsync_UsingTaskReturn_Success()` test method to test `PingProcess.RunAsync()` using `"localhost"` **without** using async/await. ✔
   - Also implement the `async public Task RunAsync_UsingTpl_Success()` test method to test `PingProcess.RunAsync()` using `"localhost"` but this time **DO** using async/await. ✔
3. Add support for an optional cancellation token to the `PingProcess.RunAsync()` signature. ✔
   Inside the `PingProcess.RunAsync()` invoke the token's `ThrowIfCancellationRequested()` method so an exception is thrown. ✔
   Test that, when cancelled from the test method, the exception thrown is an `AggregateException` ✔ with a `TaskCanceledException` inner exception ✔ using the test methods `RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrapping` ✔and `RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrappingTaskCanceledException` ✔ respectively.
4. Complete/fix **AND test** `async public Task<PingResult> RunAsync(IEnumerable<string> hostNameOrAddresses, CancellationToken cancellationToken = default)` which executes ping for an array of hostNameOrAddresses (which can all be "localhost") **in parallel**, adding synchronization if needed. ❌
  - **Not tested**

   NOTE:
      - The order of the items in the stdOutput is irrelevant and expected to be intermingled.
      - StdOutput must have all the ping output returned (no lines can be missing) even though intermingled. ✔
6. Implement **AND test** `public Task<int> RunLongRunningAsync(ProcessStartInfo startInfo, Action<string?>? progressOutput, Action<string?>? progressError, CancellationToken token)` using `Task.Factory.StartNew()` and invoking `RunProcessInternal` with a `TaskCreation` value of `TaskCreationOptions.LongRunning` and a `TaskScheduler` value of `TaskScheduler.Current`. Returning a `Task<PingResult>` is also okay.
   NOTE: This method does **NOT** use `Task.Run`. ❌
  - **Run method is used**

## Extra Credit

- Test and implement `PingProcess.RunAsync(System.IProgress<T> progess)` so that you can capture the output as it occurs rather than capturing the output only when the process finishes. ❌

## Fundamentals

- Place all shared project properties into a `Directory.Build.Props` file. 
  - **Duplicate properties in other csproj files should be removed**
- Place all shared project items into a `Directory.Build.targets` file.
- Ensure nullable reference types is enabled  ✔
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers)  ✔
- Set `LangVersion` and the `TargetFramework` to the latest released versions available (preview versions optional)   ✔
- and enabled .NET analyzers for both projects ✔
- For this assignment, consider using `Assert.AreEqual<T>()` (the generic version)  ✔
- All of the above should be unit tested ✔
- Choose simplicity over complexity ✔

**Great job overall! A couple minor tweaks to make but other than that you guys did really well.**

File: Assignment/Assignment/PingProcess.cs (hunk 4 above)
Comment (JessicaH46) [reply Assignment/Assignment/PingProcess.cs]: Thank you! Your review was very helpful and I think we were able to fix it!

File: Assignment/Assignment.Tests/PingProcessTests.cs (hunk 2 above)
Comment (JessicaH46) [reply Assignment/Assignment.Tests/PingProcessTests.cs]: thank you for the suggestion, it does provide more readability!

**Pull request closed without merging**